### DISCUSSION TITLE
Range operator ".." in multidimensional arrays

### DISCUSSION THREAD

**Discussion by erikthysell** [dotnet/csharplang#7015] [Language Ideas]
I would love for the range operator to work for multidimensional arrays, as it works in matlab, python/numpy and fortran so that for example I could do this:

(extract a x-y plane from the the x-y-z 3-d array given a index for the z-dimension (height))

```
public double[,] ExtractPlane(double[,,] threeDArray, int heightIndex)
{
    return threeDArray[..,..,heightIndex];
}

```

**Comment by svick** [comment]
Note that to some degree, you can do this yourself. For example, you could write an extension method with this signature:

```c#
public static T[,] Extract<T>(this T[,,] threeDArray, Range range1, Range range2, int index3)
```

And then use it like this:

```c#
threeDArray.Extract(.., .., heightIndex)
```

Supporting other combinations of ranges and indexes is a matter of adding more overloads of the method.

**Reply by TahirAhmadov** [reply]
Also, with #5497 and extension properties, a new indexed `this` can be added, making it work just like the OP specified.

**Comment by FaustVX** [comment]
You can use this nuget package [CommunityToolkit.HighPerformance](https://www.nuget.org/packages/CommunityToolkit.HighPerformance) made by Microsoft, which add [`Span2D<T>`](https://learn.microsoft.com/en-us/dotnet/communitytoolkit/high-performance/span2d) (and others classes), then you can `Slice` on it.

**Comment by BriannaSensf** [comment]
Unfortunately, the range operator .. in C# does not support multidimensional arrays directly. You would need to create a new 2D array and copy the values from the corresponding plane of the 3D array.

Here's an example implementation:
public double[,] ExtractPlane(double[,,] threeDArray, int heightIndex)
{
    int rows = threeDArray.GetLength(0);
    int cols = threeDArray.GetLength(1);
    double[,] plane = new double[rows, cols];
    
    for (int i = 0; i < rows; i++)
    {
        for (int j = 0; j < cols; j++)
        {
            plane[i, j] = threeDArray[i, j, heightIndex];
        }
    }
    
    return plane;
}

Alternatively, you could use the LINQ Select method to extract the plane as a 1D array, and then reshape it into a 2D array using the Buffer.BlockCopy method:
public double[,] ExtractPlane(double[,,] threeDArray, int heightIndex)
{
    int rows = threeDArray.GetLength(0);
    int cols = threeDArray.GetLength(1);
    double[] plane = threeDArray.Cast<double>()
                                 .Where((val, idx) => idx % (rows * cols) == heightIndex)
                                 .ToArray();
    
    double[,] result = new double[rows, cols];
    Buffer.BlockCopy(plane, 0, result, 0, rows * cols * sizeof(double));
    
    return result;
}
Note that this implementation assumes that the 3D array is stored in row-major order. If it's stored in column-major order, you would need to modify the code accordingly.
